compiler/modules/comm.lib/decimal.bal

```diff
@@ -0,0 +1,91 @@
+public function toLeDpd(decimal val) returns int[2] {
+    string stringRep = val.toString();
+    // this assumes toString generates a valid string representation ["+"|"-"]d+["."d+]["E"("+"|"-")d+]
+    int sign = stringRep[0] == "-" ? (1 << 63) : 0;
+    int? dot = stringRep.indexOf(".");
+    int sFirst = stringRep[0] is "+"|"-" ? 1 : 0; // index of first digit of significand
+    int? exponentStartIndex = stringRep.indexOf("E");
+    int sLast = exponentStartIndex is () ? stringRep.length() - 1 : exponentStartIndex - 1; // index of last digit of significand
+    int exponent = exponentStartIndex is () ? 0 : checkpanic int:fromString(stringRep.substring(exponentStartIndex + 1));
+    if dot !is () { // we have a dot in significand
+        exponent -= (sLast - dot);
+    }
+    byte[] significandBcd = from int i in sFirst ... sLast where i != dot select <byte>checkpanic int:fromString(stringRep[i]);
+    while exponent - 6111 > 0 {
+        // exponent is too large for exponent field
+        significandBcd.push(0);
+        exponent -= 1;
+    }
+    if significandBcd.length() > 34 {
+        panic error("overflowing significand"); // this should not happen since we are starting from a decimal
+    }
+    int paddingLen = 34 - significandBcd.length();
+    if paddingLen > 0 {
+        byte[] paddedSignificant = from int i in 0 ..< paddingLen select 0;
+        paddedSignificant.push(...significandBcd);
+        significandBcd = paddedSignificant;
+    }
+    int:Unsigned16[] dpds = from int index in  0 ... 10 select bcd3ToDeclet(significandBcd, (index * 3) + 1);
+    int mostSignificant = sign | dpdCombinationBits(exponent, significandBcd) | (dpds[0] << 36) | (dpds[1] << 26) | (dpds[2] << 16) | (dpds[3] << 6) | (dpds[4] >> 4);
+    int lestSignificant = (dpds[4] & 0xf) << 60 | (dpds[5] << 50) | (dpds[6] << 40) | (dpds[7] << 30) | (dpds[8] << 20) | (dpds[9] << 10) | dpds[10];
+    return [lestSignificant, mostSignificant];
+}
+
+// Takes 3 consecutive decimal numbers [0-9] (starting from starting index) in the significandBcd and encode them to a 10-bit declet
+// This assumes significandBcd contains 34 numbers (padded with 0)
+function bcd3ToDeclet(byte[] significandBcd, int startingIndex) returns int:Unsigned16 {
```

Comment:
Can't we take `byte[] bcds` (where `bcds.length() == 3`) as the input? You may use `array:slice` to slice `significandBcd` in the caller. Then we don't need above comment.

Reply:
fixed